cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -1620,52 +1662,76 @@ void reader::impl::allocate_columns(size_t skip_rows, size_t num_rows, bool uses
 
   // compute output column sizes by examining the pages of the -input- columns
   if (has_lists) {
-    auto& page_keys  = _chunk_itm_data.page_keys;
     auto& page_index = _chunk_itm_data.page_index;
-    for (size_t idx = 0; idx < _input_columns.size(); idx++) {
-      auto const& input_col = _input_columns[idx];
-      auto src_col_schema   = input_col.schema_idx;
-      size_t max_depth      = input_col.nesting_depth();
 
-      auto* cols = &_output_buffers;
-      for (size_t l_idx = 0; l_idx < input_col.nesting_depth(); l_idx++) {
+    std::vector<input_col_info> h_cols_info;
+    h_cols_info.reserve(_input_columns.size());
+    std::transform(_input_columns.cbegin(),
+                   _input_columns.cend(),
+                   std::back_inserter(h_cols_info),
+                   [](auto& col) -> input_col_info {
+                     return {col.schema_idx, static_cast<size_type>(col.nesting_depth())};
+                   });
+
+    auto const max_depth =
+      (*std::max_element(h_cols_info.cbegin(),
+                         h_cols_info.cend(),
+                         [](auto& l, auto& r) { return l._nesting_depth < r._nesting_depth; }))
+        ._nesting_depth;
+
+    auto const d_cols_info = cudf::detail::make_device_uvector_async(
+      h_cols_info, _stream, rmm::mr::get_current_device_resource());
+
+    // size iterator. indexes pages by sorted order
+    auto const size_input = cudf::detail::make_counting_transform_iterator(
+      0,
+      get_page_nesting_size{
+        d_cols_info.data(), max_depth, pages.size(), pages.device_ptr(), page_index.begin()});
```

Comment:
Transform iterator is known to cause inflated compiling time when used in `thrust` APIs like `reduce_by_key` and `scan_by_key`. 

In addition, `size_input` is used twice, and each time it is used the function `get_page_nesting_size::operator(index)` may be called multiple times for each input index. That is inefficient.

Please materialize it instead:
```
auto size_input = rmm::device_uvector<size_type>(pages.size(), _stream);
thrust::transform(thrust::make_counting_iterator(0), thrust::make_counting_iterator(pages.size()) , size_input.begin(), get_page_nesting_size{
        d_cols_info.data(), max_depth, pages.size(), pages.device_ptr(), page_index.begin()});

...
// use it: size_input.begin()
```

Reply:
Hopefully, this change won't bring any runtime surprises.

Reply:
@SrikarVanavasam Loos like this is the last comment to address before we can merge.